File: Services/MoonCalculator.cs
using System;

namespace SunTime.Services;

public static class MoonCalculator
{
    public record MoonData(
        double PhaseFraction,        // 0..1  (0=new, 0.25=first quarter, 0.5=full, 0.75=last quarter)
        double IlluminatedFraction,  // 0..1
        double HourOnDial            // 0..24 — position on the 24 h clock ring
    );

    // Known new moon: 2000-01-06 18:14 UTC
    private const double NewMoonJd     = 2451550.26;
    private const double SynodicPeriod = 29.530588853; // days

    public static MoonData Calculate(DateTime utcNow, double localHour)
    {
        double jd      = ToJulianDay(utcNow);
        double ageDays = ((jd - NewMoonJd) % SynodicPeriod + SynodicPeriod) % SynodicPeriod;
        double fraction    = ageDays / SynodicPeriod;
        double illuminated = (1.0 - Math.Cos(fraction * 2.0 * Math.PI)) / 2.0;

        // Moon falls behind the sun by one full 24 h revolution per synodic month.
        // localHour anchors the moon in the same coordinate system as the sun icon.
        double hourOnDial = (localHour + fraction * 24.0) % 24.0;

        return new MoonData(fraction, illuminated, hourOnDial);
    }

    private static double ToJulianDay(DateTime utc)
    {
        int    y = utc.Year;
        int    m = utc.Month;
        double d = utc.Day + utc.TimeOfDay.TotalDays;
        if (m <= 2) { y--; m += 12; }
        int A = y / 100;
        int B = 2 - A + A / 4;
        return (int)(365.25 * (y + 4716)) + (int)(30.6001 * (m + 1)) + d + B - 1524.5;
    }
}

File: apple-watch/SunTimeWatch Watch App/MoonCalculator.swift
import Foundation

enum MoonCalculator {
    private static let newMoonJD     = 2451550.26
    private static let synodicPeriod = 29.530588853

    static func calculate(utcNow: Date, localHour: Double) -> MoonData {
        let jd       = julianDay(utcNow)
        let ageDays  = fmod(fmod(jd - newMoonJD, synodicPeriod) + synodicPeriod, synodicPeriod)
        let fraction = ageDays / synodicPeriod
        let illum    = (1.0 - cos(fraction * 2.0 * .pi)) / 2.0
        let hourOnDial = fmod(localHour + fraction * 24.0, 24.0)
        return MoonData(phaseFraction: fraction, illuminatedFraction: illum, hourOnDial: hourOnDial)
    }

    private static func julianDay(_ date: Date) -> Double {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let c = cal.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)
        var y = c.year!; var m = c.month!
        let d = Double(c.day!) + (Double(c.hour!) * 3600 + Double(c.minute!) * 60 + Double(c.second!)) / 86400.0
        if m <= 2 { y -= 1; m += 12 }
        let A = y / 100; let B = 2 - A + A / 4
        return Double(Int(365.25 * Double(y + 4716))) + Double(Int(30.6001 * Double(m + 1))) + d + Double(B) - 1524.5
    }
}
